File: Math/007.Reverse-Integer/007.Reverse-Integer.java
class Solution {
    public int reverse(int x) {
        // 最小值取绝对值会溢出，我们需要先判断一下是不是最小值
        if (x == Integer.MIN_VALUE) {
            return 0;
        }

        int sign = 1;
        if (x < 0) {
            sign = -1;
        }

        x = Math.abs(x);

        int ret = 0;
        while (x > 0) {
            // 这里也是很巧妙，我们如何在不溢出的情况下去判断 ret * 10 > MAX_VALUE 呢
            // 转换成 ret > Integer.MAX_VALUE / 10 就好了
            if (ret > Integer.MAX_VALUE / 10) {
                return 0;
            }
            // 与上同
            if (ret * 10 > Integer.MAX_VALUE - x % 10) {
                return 0;
            }

            ret = ret * 10 + x % 10;
            x = x / 10;
        }
        return sign * ret;
    }
}

File: Math/007.Reverse-Integer/007.Reverse-Integer.ts
function reverse(x: number): number {
  const numMax = 2 ** 31;
  const numMin = -(2 ** 31 - 1);
  let num = x;

  const isNegative = num < 0;
  num = Math.abs(num);
  let ret = 0;
  while (num > 0) {
    ret = ret * 10 + (num % 10);
    num = Math.floor(num / 10);
  }
  ret = isNegative ? -ret : ret;
  const isOutOfRange = ret > numMax || ret < numMin;
  return isOutOfRange ? 0 : ret;
}
